### DISCUSSION TITLE
When using cursor pagination, given an item's ID, how to fetch the previous and next items?

### DISCUSSION THREAD

**Discussion by GP4cK** [TriPSs/nestjs-query#133] [Q&A]
Let's say I have a paginated list of `items` which are displayed in a table when I go to `/items`.
When I click on an item, it should take me to `/item/:id`.
On an item page, there are arrows which allow me to quickly go to the previous or the next item without going back to the table view.

If a user copy/pastes the url `/item/:id` in the browser, I wouldn't know what is it's cursor. How can I fetch the next and previous items?

**Comment by TriPSs** [comment]
Interesting, this is a case that I'm also looking at in our application so I'm very interested on how to properly solve this. 

Question: Is it in your use case also possible for the user to filter the table? If so should that be reflected back into the next/previous items?

**Reply by GP4cK** [reply]
Yes, they could filter or sort which will affect the ordering.

**Reply by TriPSs** [reply]
Then we have almost the same use case here, the only difference is that we are not using a cursor but offset further the idea is the same.

**Reply by GP4cK** [reply]
I was thinking to avoid offset as I read in a few places that it doesn't scale well (including in the documentation of this repo).
The best article I found so far is this one but it doesn't fully cover our use case: https://apryse.com/blog/graphql/implementing-graphql-pagination

From my research and discussions with ChatGPT, I see 2 options:
1. query all the items (at least the id). Then it's easy to know what is the index of an item in the list but it defeats the purpose of the pagination.
2. rebuild the cursor on the frontend and make one query to fetch the previous item and one query to fetch the next item. Here the issue is that it couples the implementation of the frontend and backend, and, if the item hasn't been returned in a previous `items` query, you will not know what is its index. For reference, [here](https://github.com/TriPSs/nestjs-query/blob/1dbbe5c9c1299bfb729559dfcc3fbed22988d56e/packages/query-graphql/src/types/connection/cursor/pager/strategies/keyset.pager-strategy.ts#L30-L33) is where the cursor is created for KeySet pagination. It depends on the KeySet and the sort parameters.

**Reply by GP4cK** [reply]
A third option would be to do something like Jira and cap the number of results which can be returned for a given filter at 1,000. If you want to know what's after, you need a more precise filter.
